### DISCUSSION TITLE
saned is not show Airscan scanners

### DISCUSSION THREAD

**Discussion by zekzit** [alexpevzner/sane-airscan#319] [General]
I have use `scanimage -L` it show devices that prefix with `airscan:<scanner>`. But when I let `scanimage` listing through my local saned, it show only device that prefix with `net:localhost:<scanner>` but not `net:localhost:airscan:<scanner>`. So, how can I share this airscan scanner through saned? Or it cannot be shared?

**Comment by alexpevzner** [comment]
Hi @zekzit,

this is by design of `saned`. When SANE "frontend" asks its "backends" for a list of connected devices, using the `sane_get_devices` function, it can specify via its second parameter, if it is interested in all devices or locally connected only ones.

As eSCL/WSD devices are all network devices, i.e. not locally-connected, `sane-airscan` returns empty list when only local devices are requested.

I've briefly reviewed another backends, they behave on a similar manner.

I wonder why somebody may want to re-export network device via `saned` instead of connecting to it directly.

But anyway, if you feel that `saned` behaviour should be somehow changed, it is better to file issue into the saned repository at GitLab: https://gitlab.com/sane-project/backends

**Comment by alexpevzner** [comment]
I will convert this conversation to discussion, as it is not sane-airscan issue actually...

**Comment by zekzit** [comment]
Hi @alexpevzner,

First of all, thanks for answer my question. I am a Java developer that want to add new scanner to my application. This application is rely on Morena 6 library that using saned to connect with scanner. The easiest way to have it work right away is to config saned to re-export network device to let my application connect to (which is a little bit detour) so I asked if I can do it this way or not and you have answer me clearly.

Now I have to consider 2 more ways which are writing a wrapper that call `scanimage` CLI or finding a way to connect Airscan/eSCL device directly. But in either way, it has to take some time to do it.

In my opinion, `saned` behavior is OK. But it might be good if it have option/parameter to let us re-export network device.

**Comment by alexpevzner** [comment]
Hi @zekzit,

saned protocol is specific to SANE, while eSCL is industry standard, now published, and many devices support it directly. There are also AirSane project (https://github.com/SimulPiscator/AirSane) able to share locally connected scanners using eSCL protocol.

Did you ever consider implementing eSCL in your Java app? I think it has many advantages over implementing the saned protocol...

**Comment by gutschke** [comment]
I recently ran into the same issue. I fully understand the author's preference to leave the current behavior alone. In a perfect world, that's certainly the correct behavior. But sometimes, we have to deal with broken systems that are impossible to fix -- and it turns out that in my case, `sane-airscan` was the least-offensive component to patch.

I wish this behavior was documented somewhere prominently. Took me way too many hours to figure out why I couldn't scan in some of my programs, and to come up with a work-around. I am sure there are other users out there who simply just give up. In the interest of helping everyone who encounters the same frustrating behavior and is OK with patching their local copy of `sane-airscan`, I am attaching a "diff" file.
[override-local.diff.txt](https://github.com/user-attachments/files/18534175/override-local.diff.txt)

Of course if @alexpevzner decided to apply this patch as an ugly but pragmatic solution to a nasty edge-case, that would be wonderful. One less patch that I need to carry locally.

**Comment by alexpevzner** [comment]
Hi @gutschke,

please note, nobody so far has manages to send PR against `sane-airscan` that adds this option.

If this PR will be created, I will consider merging it into the driver source. Please note, man page and the sample config file needs to be updated as well.

Regarding option name, I would prefer to call it "pretend-local", not "override" (and with dash, not underscore, for consistency with other multi-word option names, like ws-discovery)

**Reply by gutschke** [reply]
I just sent you a pull request and incorporated the changes that you asked for. Feel free to edit any way that you see fit, or send me comments and I'll make the changes on my end.
